File: frontend/test_api.py
import httpx
import time

def run_tests():
    print("=== Testing Backend Endpoints ===")
    
    # 1. Upload
    print("1. Uploading image to /api/v1/upload...")
    with open("data/demo/image.jpg", "rb") as f:
        files = {"file": ("image.jpg", f, "image/jpeg")}
        upload_res = httpx.post("http://localhost:8000/api/v1/upload", files=files)
    
    upload_data = upload_res.json()
    image_id = upload_data.get("image_id")
    print(f"-> Upload Response Image ID (UUID): {image_id}")
    
    if not image_id:
        print("Failed to get image_id")
        return

    def submit_query(query):
        print(f"\n-> Submitting query: '{query}' with image_id: {image_id}")
        query_res = httpx.post("http://localhost:8000/api/v1/query", json={
            "query": query,
            "image_ids": [image_id],
            "include_evidence": True,
            "include_trace": True
        })
        
        query_data = query_res.json()
        task_id = query_data.get("task_id")
        if task_id:
            status = "processing"
            while status not in ["completed", "failed"]:
                time.sleep(1)
                stat_res = httpx.get(f"http://localhost:8000/api/v1/status/{task_id}")
                status = stat_res.json().get("status")
                
            if status == "completed":
                res_res = httpx.get(f"http://localhost:8000/api/v1/result/{task_id}")
                res_data = res_res.json().get("result", {})
                print(f"-> Final Answer: {res_data.get('answer')}")
                print(f"-> Confidence: {res_data.get('confidence', {}).get('score')}")
                boxes = [e for e in res_data.get('evidence', []) if e.get('type') == 'bounding_box']
                if boxes:
                    print(f"-> Found {len(boxes)} bounding box(es)")
                    for b in boxes:
                        print(f"   Box label: {b.get('data', {}).get('label')}")
            else:
                print("-> Task failed.")
        else:
            print("-> Query Error:", query_data)

    print("\n--- Testing VQA ---")
    submit_query("What objects are visible in this image?")
    submit_query("Is there a basketball court?")

    print("\n--- Testing Grounding ---")
    submit_query("Where are the buildings?")
    submit_query("Locate the basketball court.")
    submit_query("Locate the road.")

if __name__ == "__main__":
    run_tests()

File: frontend/test_api.mjs
import fs from 'fs';

async function runTests() {
    console.log("=== Testing Backend Endpoints ===");
    
    // 1. Upload a dummy image to verify backend returns a UUID
    console.log("1. Uploading image to /api/v1/upload...");
    const form = new FormData();
    const dummyFile = new Blob(["dummy data for testing"], { type: "image/tiff" });
    form.append("file", dummyFile, "test_image.tif");
    
    const uploadRes = await fetch("http://localhost:8000/api/v1/upload", {
        method: "POST",
        body: form
    });
    
    const uploadData = await uploadRes.json();
    console.log("-> Upload Response Image ID (UUID):", uploadData.image_id);
    
    const imageId = uploadData.image_id;

    async function submitQuery(query) {
        console.log(`\n-> Submitting query: "${query}" with image_id: ${imageId}`);
        const queryRes = await fetch("http://localhost:8000/api/v1/query", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                query: query,
                image_ids: [imageId],
                include_evidence: true,
                include_trace: true
            })
        });
        
        const queryData = await queryRes.json();
        if (queryData.task_id) {
            let status = "processing";
            while (status !== "completed" && status !== "failed") {
                await new Promise(r => setTimeout(r, 1000));
                const statRes = await fetch(`http://localhost:8000/api/v1/status/${queryData.task_id}`);
                const statData = await statRes.json();
                status = statData.status;
            }
            if (status === "completed") {
                const resRes = await fetch(`http://localhost:8000/api/v1/result/${queryData.task_id}`);
                const resData = await resRes.json();
                console.log("-> Final Answer:", resData.result.answer);
                console.log("-> Confidence:", resData.result.confidence.score);
                // Check if any boxes
                const boxes = resData.result.evidence.filter(e => e.type === "bounding_box");
                if (boxes.length > 0) {
                    console.log(`-> Found ${boxes.length} bounding box(es)`);
                    boxes.forEach(b => {
                        console.log(`   Box label: ${b.data.label}`);
                    });
                }
            } else {
                console.log("-> Task failed.");
            }
        } else {
            console.log("-> Query Error:", queryData);
        }
    }

    // 2. Submit VQA queries
    console.log("\n--- Testing VQA ---");
    await submitQuery("What objects are visible in this image?");
    await submitQuery("Is there a basketball court?");

    // 3. Submit Grounding queries
    console.log("\n--- Testing Grounding ---");
    await submitQuery("Where are the buildings?");
    await submitQuery("Locate the basketball court.");
    await submitQuery("Locate the road.");

}

runTests().catch(console.error);
